### DISCUSSION TITLE
内存消耗简单测试 [macOS]

### DISCUSSION THREAD

**Discussion by hinego** [xjasonlyu/tun2socks#109] [Ideas]
开启tun2socks后 大概15分钟

中间看过一点youtube

用chome无痕模式开了四个speedtest.com

期间最大内存使用87MB

关闭无痕模式后使用65.3MB

关闭Crome后使用27.6MB

再反复一遍 内存使用 67Mb

关闭chorme后接近2分钟内无变化

关闭chorme后接近4分钟左右后降低到41.7Mb

关闭chrome后接近10分钟后 降低到20.6MB

使用的程序版本为：https://github.com/xjasonlyu/tun2socks/commit/b166ed5e66c100d39c1b3f5938181953053217dd

**Comment by hinego** [comment]
另外的版本 较稳定

开七个无痕speedtest

稳定在26MB左右

关掉chomre一段时间后降低到9MB

**Reply by xjasonlyu** [reply]
另外的版本是指其他的tun2socks实现吗

**Reply by hinego** [reply]
> 另外的版本是指其他的tun2socks实现吗

https://github.com/eycorsican/go-tun2socks

这个版本

你那版我没编译成功后续可以再试试

**Comment by hinego** [comment]
https://github.com/eycorsican/go-tun2socks 这个版本

挂机自动刷新google 页面

经过一段时间后崩了 go routine未释放

`
goroutine 395817 [IO wait, 3 minutes]:
`

错误大概如上

**Comment by hinego** [comment]
再测试 https://github.com/eycorsican/leaf

启动瞬间内存：2~3M左右

使用一段时间后：7.6M （speedtest+youtube观看+gmail等）
内存一直很稳定 7.6M

关闭chrome后5分钟：7.6M（貌似它会一直保持这个内存，留作缓存，后续有一定新连接时也不会新增内存）

但是可惜不会rust 大多数协议目前都只有 golang版本


也可惜原生go实现的 内存消耗太大了

**Comment by xjasonlyu** [comment]
你有兴趣可以跑一下带宽测试，我之前用go-tun2socks 最多只能跑到100Mbps，不清楚leaf如何。

**Reply by hinego** [reply]
怎么本地测试啊 没找到较好的测试方式

。。。

**Reply by hinego** [reply]
但是内存是真的稳 上的7.6MB后稳了一上午 未有增加 速度应该也不错的

**Reply by xjasonlyu** [reply]
对于现代设备来说，几十MB的内存占用是完全没有感觉的

**Reply by hinego** [reply]
确实是这么个理

但是IOS network extension

ios15前限制15M
ios15后限制50M

15M跑纯go版本就是痴人说梦了

跑cgo版本也基本不太现实

rust版本完全没问题

50M跑纯go版本 也很难

跑cgo版本倒是问题不大

跑rust那就搓搓有余了

**Reply by xjasonlyu** [reply]
不过这个项目本身也就没考虑在ios上用，他的很多地方也都是针对linux做优化的

**Comment by xjasonlyu** [comment]
带宽测试方法：

1. 安装iperf3
2. 主机A启动iperf服务端
3. 主机B启动tun2socks
4. 主机B `ip route add <A IP> dev tun0`
5. 主机B `iperf3 -c <A IP>`

**Comment by hinego** [comment]
重试了一下cgo版本的 发现go routine 数量没问题（开了pprof）

猜测可能是昨晚mac休眠后产生了BUG

**Comment by hinego** [comment]
突然想起来 你说tun2socks-go(cgo)版本只能跑 100Mbps

肯定有问题

v2rayN 什么等一些列好多软件（特别是移动客户端）都是基于cgo版本的 速度还是很快的（几百M很常见）

我测速折腾了一下 没搞定

**Reply by xjasonlyu** [reply]
不知道客户端有没有针对优化过，我当时自己测确实只有100Mbps，但我没有细究

**Comment by xjasonlyu** [comment]
在最新的commit里新增了 tcp-sndbuf 和 tcp-rcvbuf 两个参数，在一些内存较小的机器里可以使用如 `-tcp-sndbuf 32k` 和 `-tcp-rcvbuf 32k`来降低内存使用量。

实测效果不错，在一些高并发高连接测速中，内存使用量足足降低了80%多。

PS: 过低的缓冲区大小会导致较高的CPU消耗，所以这是一个取舍的过程。

**Reply by xjasonlyu** [reply]
@hinego 你可以试一下新的参数，设置在32-128k之间，内存消耗应该会降低很多。
